### PR TITLE
revsync support

### PR DESCRIPTION
**Pull request by atlas0fd00m**
make each parser add a sha256 hash of the file loaded as FileMetadata
make Vivisect Extensions honor plugins as directories, not just .py files (__init__.py still will have vivExtension(vw, vwgui) called)

### CODE REVIEW

File: vivisect/parsers/blob.py [hunk 1]
```diff
@@ -31,6 +31,7 @@ def parseFd(vw, fd, filename=None, baseaddr=None):
     vw.setMeta('DefaultCall', archcalls.get(arch,'unknown'))
 
     bytez =  fd.read() 
+    vw.setFileMeta('sha256', v_parsers.sha256Bytes(bytez))
```
Comment (rakuy0) [inline vivisect/parsers/blob.py:34]: Also need the filename here as the first parameter (that's what the tests are failing on)

File: vivisect/parsers/elf.py [hunk 2]
```diff
@@ -164,15 +164,23 @@ def loadElfIntoWorkspace(vw, elf, filename=None, baseaddr=None):
...
+        for dyn in elf.getDynamics():
+            if dyn.d_tag == Elf.DT_SONAME:
+                filename = dyn.getName()
 
-    fhash = "unknown hash"
-    if os.path.exists(filename):
-        fhash = v_parsers.md5File(filename)
+        # if all else fails, fallback
+        if filename is None:
+            filename = "elf_%.8x" % baseaddr
```
Comment (rakuy0) [inline vivisect/parsers/elf.py:180]: Does this have any rippling effects or is it not hit?

File: vivisect/parsers/pe.py [hunk 3]
```diff
@@ -115,11 +115,15 @@ def loadPeIntoWorkspace(vw, pe, filename=None, baseaddr=None):
...
-    fhash = "unknown hash"
-    if os.path.exists(filename):
-        fhash = v_parsers.md5File(filename)
+    # grab the file bytes for hashing
+    pe.fd.seek(0)
+    bytez = pe.fd.read()
+    fhash = v_parsers.md5Bytes(bytez)
+    sha256 = v_parsers.sha256Bytes(bytez)
 
+    # create the file and
```
Comment (rakuy0) [inline vivisect/parsers/pe.py:124]: Just to be persnickety, can you finish the comment?

File: vivisect/parsers/elf.py (hunk 2 above)
Comment (atlas0fd00m) [reply vivisect/parsers/elf.py:180]: this is a last-ditch effort to not crash or have no filename.

File: vivisect/parsers/blob.py (hunk 1 above)
Comment (atlas0fd00m) [reply vivisect/parsers/blob.py]: yup, <facepalm>

File: vivisect/parsers/pe.py (hunk 3 above)
Comment (atlas0fd00m) [reply vivisect/parsers/pe.py]: yup :)

File: vivisect/extensions/__init__.py [hunk 4]
```diff
@@ -28,19 +28,25 @@ def loadExtensions( vw, vwgui ):
...
+            # otherwise, run all the .py files in the VIV_EXT_PATH dir
+            if not modpath.endswith('.py'):
                 continue
 
             # Build code objects from the module files
             mod = imp.new_module('viv_ext')
-            filepath = os.path.join(dirname, fname)
-            filebytes = file( filepath, 'r' ).read()
-            mod.__file__ = filepath
+            filebytes = file( modpath, 'r' ).read()
```
Comment (rakuy0) [inline vivisect/extensions/__init__.py:44]: Please don't use `file`. I know it works in python 2, but it got removed as a keyword in python 3.

Comment (atlas0fd00m) [reply vivisect/extensions/__init__.py:44]: lol, that wasn't something i changed.
fixing anyway.

**Pull request merged**